lib/typerb.c

```diff
@@ -376,10 +377,10 @@ struct typed_rb_entry *typed_rb_insert(struct rbt_tree *rbt,
 }
 
 /* Finds the node with the same key as elm */
-struct rb_entry *typed_rb_find(struct rbt_tree *rbt, const struct rb_entry *key,
-		int (*cmpfn)(
-			const struct typed_rb_entry *a,
-			const struct typed_rb_entry *b))
+struct rb_entry *typed_rb_find(const struct rbt_tree *rbt,
```

Comment:
NAK on returning a non-const `rb_entry` from a const `rbt_tree`

As much as I like putting `const` on things, this is a situation where it doesn't really work out.

Reply:
Sorry, can you elaborate a bit more on why this is an issue? Is actually a code issue or just an API nit because I can't produce a test case where it is breaking stuff.

Also, is it just a problem with the `_find()` api or do `_first()` and others need to be changed back as well?

Reply:
The problem is that `const` on the data structure head doesn't have clearly defined semantics.

* does it mean that I'm not supposed to allocate/free the head?  That doesn't work because the DS might need to change something in the head on insertion, so cross that out...
* does it mean that I'm not supposed to change the DS but can do with the items what I want?  That's kinda weird... why is the DS `const` but not the items?
* does it mean that I'm not supposed to change anything?  Then the items need to be const too...

Even if we agree on a semantic to attach, I'd argue that this is non-obvious enough to make sticking `const` on the DS head useless.

(My comment applies to all functions, `_first` & co too.)

So — what's your use case?

Reply:
I didn't think this was a problem when I saw it - though I think I understand your question. The use-case for `const` collection but not `const` item would be that I want to modify an item in a collection, but not the collection itself.

Reply:
ok, yea that makes sense. Thanks for taking the time to explain that.

I mostly decided whether my const decisions were appropriate by looking at the linux implementation: https://github.com/torvalds/linux/blob/master/include/linux/rbtree.h

But I definitely agree it in non-obvious, so I will remove it.

I will need to remove it in the other API's in `typesafe.h` as well,

Reply:
@eqvinox and @sworleys -> does any of this conversation need to be recorded in doc/developer somewhere?